pkg/appfile/template.go

```diff
@@ -73,120 +77,187 @@ func GetScopeGVK(ctx context.Context, cli client.Reader, dm discoverymapper.Disc
+					}
+					tmpl, err := newTemplateOfCompDefinition(compDef)
+					if err != nil {
+						return nil, errors.WithMessagef(err, "cannot load template of component definition %q", capName)
+					}
+					return tmpl, nil
+				}
+				if unstructDef.GetKind() == v1beta1.TraitDefinitionKind &&
+					capType == types.TypeTrait && unstructDef.GetName() == capName {
+					traitDef := &v1beta1.TraitDefinition{}
+					if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructDef.Object, traitDef); err != nil {
+						return nil, errors.Wrap(err, "invalid trait definition")
+					}
+					tmpl, err := newTemplateOfTraitDefinition(traitDef)
+					if err != nil {
+						return nil, errors.WithMessagef(err, "cannot load template of trait definition %q", capName)
+					}
+					return tmpl, nil
+				}
+				// TODO(roywang) add support for ScopeDefinition
+			}
+		}
+		// not found in provided cap definitions
+		// then try to retrieve from cluster
+		tmpl, err := LoadTemplate(ctx, dm, r, capName, capType)
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return nil, fmt.Errorf("cannot load template %q from cluster and provided ones", capName)
+			}
+			return nil, errors.WithMessage(err, "cannot load template")
+		}
+		return tmpl, nil
+	})
+}
 
-	if status != nil {
-		tmp.CustomStatus = status.CustomStatus
-		tmp.Health = status.HealthPolicy
+func newTemplateOfCompDefinition(compDef *v1beta1.ComponentDefinition) (*Template, error) {
+	tmpl := &Template{
+		Reference:           compDef.Spec.Workload.Definition,
+		ComponentDefinition: compDef,
+	}
+	if err := loadSchematicToTemplate(tmpl, compDef.Spec.Status, compDef.Spec.Schematic, compDef.Spec.Extension); err != nil {
+		return nil, errors.WithMessage(err, "cannot load template")
 	}
-	if schematic != nil {
-		if schematic.CUE != nil {
-			tmp.TemplateStr = schematic.CUE.Template
-			// CUE module has highest priority
-			// no need to check other schematic types
-			return tmp, nil
+	if compDef.Annotations["type"] == string(types.TerraformCategory) {
+		tmpl.CapabilityCategory = types.TerraformCategory
+	}
+	return tmpl, nil
+}
+
+func newTemplateOfTraitDefinition(traitDef *v1beta1.TraitDefinition) (*Template, error) {
+	tmpl := &Template{
+		TraitDefinition: traitDef,
+	}
+	if err := loadSchematicToTemplate(tmpl, traitDef.Spec.Status, traitDef.Spec.Schematic, traitDef.Spec.Extension); err != nil {
+		return nil, errors.WithMessage(err, "cannot load template")
+	}
+	return tmpl, nil
+}
+
+func newTemplateOfWorkloadDefinition(wlDef *v1beta1.WorkloadDefinition) (*Template, error) {
+	tmpl := &Template{
+		WorkloadDefinition: wlDef,
+	}
+	if err := loadSchematicToTemplate(tmpl, wlDef.Spec.Status, wlDef.Spec.Schematic, wlDef.Spec.Extension); err != nil {
+		return nil, errors.WithMessage(err, "cannot load template")
+	}
+	return tmpl, nil
+}
+
+func loadSchematicToTemplate(tmpl *Template, sts *common.Status, sche *common.Schematic, ext *runtime.RawExtension) error {
```

Comment:
let‘s write more  some tests for these four new functions

Reply:
sure